### DISCUSSION TITLE
Web server utilizing Async-http with a twist: Ractors!

### DISCUSSION THREAD

**Discussion by Forthoney** [ioquatix/ioquatix#32] [Show and tell]
I built an experimental [web server](https://github.com/Forthoney/mooro/tree/async) which uses Ractors as workers rather than Threads or separate processes. It utilizes [Async-http](https://github.com/socketry/protocol-http) to concurrently handle requests since Ractors are costly to spawn and having a Ractor per request would be far too costly (apache vs nginx but in ruby).

I am sharing this here because I encountered some friction points in the process of integrating `Async-http` with `Mooro`. I would love to work together to address these points and make `Async-http` and related gems easier to use for not just my project, but for other people looking into using `Async-http`.

To name a couple that come to mind right now,
### [Protocol-rack](https://github.com/socketry/protocol-rack) hardcodes the use of [console](https://github.com/socketry/console). 

I don't mind using console (imo it is indeed better than the default console), but due to the tight object sharing constraints of ractors, I need to invoke console indirectly through Ractor messages (almost like an rpc). `Protocol-rack` can be easily refactored to allow users to pass in a custom object in place of Console

### Enable shareable constant literals by default

A quick and easy QoL change would be to enable constant literals to be shareable. This just means I don't have to run tons of `Ractor.make_shareable(const)`. A simple one line magic comment `# shareable_constant_value: literal` is all you need to implement this change.

**Comment by ioquatix** [comment]
I think it's okay to remove `Console` from `protocol-rack` gem if there isn't a significant need for logging. Logging can be added after the fact. Alternatively, it might be a good idea to make `Console` gem compatible with `Ractor`.

Regarding shareable literals, I think it's okay, however I'm pursuing a different mechanism: https://bugs.ruby-lang.org/issues/18035 - maybe we can make a shim for this which does a deep freeze (similar to `Ractor.make_shareable`). `shareable_constant_value` is probably an experimental feature..?

**Reply by Forthoney** [reply]
1. Awesome! I can start make a pr for making `Console` optional. I think making `Console` itself Ractor compatible would be awesome, but this should be a more long term goal considering this will probably take a good amount of time.

2. It is experimental in the sense that Ractors are still experimental (and frankly very very buggy, but that's a different topic). But I do think they are pushing this to be the canonical way to make literals shareable. We can hold off on this while more discussions/progress is made on Ractors. Hopefully this happens sooner rather than later because the regression in Ractors in 3.2 -> 3.3 is quite bad

**Reply by ioquatix** [reply]
As a first step, please create issues on the appropriate gems, and feel free to make PRs so we can discuss the specifics.

**Reply by ioquatix** [reply]
I took a look at `console` gem and it wasn't too hard to make it support `Ractor`.

https://github.com/socketry/console/commit/e8fd32769d9051deee1c7a8b7ae6a713fe8fd393

I released it.

**Reply by ioquatix** [reply]
> Protocol-rack can be easily refactored to allow users to pass in a custom object in place of Console

By the way, anyone can already do this, e.g. `Protocol::Rack::Console = ...` is sufficient.

**Reply by Forthoney** [reply]
That's incredible. Thank you so much!
